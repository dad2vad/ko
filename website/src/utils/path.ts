export const sep = "/";

export function isAbsolute(path: string): boolean {
  return path.startsWith(sep);
}

export function isRelative(path: string): boolean {
  return path.startsWith(`.${sep}`) || path.startsWith(`../${sep}`);
}

export function dirname(path: string): string {
  const index = path.lastIndexOf(sep);
  if (index === -1) return ".";
  return path.substring(0, index) || sep;
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf(sep);
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split(sep).filter(Boolean);
  const toParts = to.split(sep).filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join(sep);
}
