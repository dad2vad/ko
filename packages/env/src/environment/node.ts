import { getRequireExtensions } from "@kosko/require";
import { join } from "node:path";
import { formatPath } from "../paths";
import { toArray } from "@kosko/common-utils";
import {
  createEnvironment,
  createLoaderReducers,
  createReducerList
} from "./base";
import { Environment, Loader, ReducerExecutor, ReducerList } from "./types";
import { cwd as getCwd } from "node:process";

export function getNodeExtensions(): string[] {
  return getRequireExtensions().map((ext) => ext.substring(1));
}

/**
 * Options for creating a Node.js environment.
 *
 * @public
 */
export interface NodeEnvironmentOptions {
  /**
   * Current working directory (CWD).
   *
   * @defaultValue `process.cwd()`
   */
  cwd?: string;
}

export interface InternalNodeEnvironmentOptions extends NodeEnvironmentOptions {
  createReducerExecutor(this: void, reducers: ReducerList): ReducerExecutor;
  requireModule(this: void, environment: Environment, id: string): any;
  mergeValues(this: void, data: any[]): any;
}

export function createNodeEnvironment({
  cwd = getCwd(),
  createReducerExecutor,
  requireModule,
  mergeValues
}: InternalNodeEnvironmentOptions): Environment {
  function requireAllEnvs(template: string, component?: string) {
    if (!environment.env) return [];

    const envs = toArray(environment.env);

    return envs.map((env) => {
      const path = formatPath(template, {
        environment: env,
        ...(component && { component })
      });

      return requireModule(environment, join(environment.cwd, path));
    });
  }

  const loader: Loader = {
    global: () => requireAllEnvs(environment.paths.global),
    component: (name) => requireAllEnvs(environment.paths.component, name)
  };
  const reducers = createReducerList(createLoaderReducers(loader, mergeValues));
  const { reduce } = createReducerExecutor(reducers);
  const environment = createEnvironment({
    cwd,
    extensions: getNodeExtensions(),
    setReducers: reducers.setReducers,
    resetReducers: reducers.resetReducers,
    reduce
  });

  return environment;
}
