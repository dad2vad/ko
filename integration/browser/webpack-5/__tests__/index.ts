import { dirname, join } from "node:path";
import execa from "execa";

const testDir = dirname(__dirname);

beforeAll(async () => {
  await execa("webpack", [], {
    cwd: testDir
  });

  await page.goto(`file://${join(testDir, "index.html")}`);
});

test("should print manifests", async () => {
  const element = await expect(page).toMatchElement("pre");
  const innerTextProp = await element.getProperty("innerText");
  const innerText = await innerTextProp.jsonValue();

  expect(innerText).toMatchSnapshot();
});
