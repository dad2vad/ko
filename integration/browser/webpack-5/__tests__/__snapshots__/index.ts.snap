// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should print manifests 1`] = `
"---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
        - image: gcr.io/image-dev/nginx:latest
          name: nginx
          ports:
            - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
spec:
  selector:
    app: nginx
  type: ClusterIP
  ports:
    - port: 80
      targetPort: 80
"
`;
